python/lsst/ts/tcpip/client.py

```diff
@@ -0,0 +1,262 @@
+        self.writer: asyncio.StreamWriter | None = None
+
+        # Task which is set done when client connects.
+        self.start_task: asyncio.Future = asyncio.create_task(self.start(**kwargs))
+
+        # Task which is set done when the client is closed.
+        self.done_task: asyncio.Future = asyncio.Future()
+
+    @property
+    def connected(self) -> bool:
+        """Return True if this client is connected to a server."""
+        return not (
+            self.reader is None
+            or self.writer is None
+            or self.reader.at_eof()
+            or self.writer.is_closing()
+        )
+
+    async def call_connect_callback(self) -> None:
+        """Call self.__connect_callback.
+
+        This is always safe to call. It only calls the callback function
+        if that function is not None and if the connection state has changed
+        since the last time this method was called.
+        """
+        connected = self.connected
+        self.log.debug(
+            f"call_connect_callback: {connected=}; was_connected={self._was_connected}"
+        )
+        if self._was_connected != connected:
+            self._was_connected = connected
+            if self.__connect_callback is not None:
+                try:
+                    await self.__connect_callback(self)
+                except Exception:
+                    self.log.exception("connect_callback failed.")
+
+    async def basic_close(self) -> None:
+        """Close the connected client socket, if any.
+
+        Like `close` except does not clear ``self.should_be_connected``,
+        nor cancel ``self._monitor_connection_task``.
+        """
+        try:
+            self.log.info("Closing")
+            if self.writer is None:
+                return
+
+            writer = self.writer
+            self.writer = None
+            await utils.close_stream_writer(writer)
+        except Exception:
+            self.log.exception("close failed; continuing")
+        finally:
+            await self.call_connect_callback()
+            if self.done_task.done():
+                self.done_task.set_result(None)
+
+    async def close(self) -> None:
+        """Close the connected client socket, if any.
+
+        Call connect_callback if a client was connected.
+        """
+        self.should_be_connected = False
+        self._monitor_connection_task.cancel()
+        await self.basic_close()
+
+    async def start(self, **kwargs: typing.Any) -> None:
+        """Connect to the TCP/IP server.
+
+        This is called automatically by the constructor,
+        and is not intended to be called by the user.
+        It is a public method so that subclasses can override it.
+
+        Parameters
+        ----------
+        **kwargs : dict[str, Any]
+            Additional keyword arguments for `asyncio.open_connection`,
+            beyond host and port.
+
+        Raises
+        ------
+        RuntimeError
```

Comment:
`RuntimeError`

Reply:
OK. I'll change all the Raises sections to match. I had not realized it was needed (I thought sphinx was smart enough to make these links without it) but you are right.